contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,156 @@
+        address _sushi,
+        address _xSushi,
+        address _treasury,
+        AllocatorInitData memory data
+    ) BaseAllocator(data) {
+        require(_chef != address(0));
+        masterChef = _chef;
+        require(_sushi != address(0));
+        sushi = _sushi;
+        require(_xSushi != address(0));
+        xSushi = _xSushi;
+        require(_treasury != address(0));
+        treasury = _treasury;
+
+        // approve for safety, sushi is being staked and xsushi is being instantly sent to treasury and that is fine
+        // but to be absolutely safe this one approve won't hurt
+        IERC20(sushi).approve(address(extender), type(uint256).max);
+        IERC20(xSushi).approve(address(extender), type(uint256).max);
+    }
+
+    /**
+     1- Deposit LP token into onsen, if deposit is succesfull we should get in return sushi tokens
+     2- Stake the Sushi rewards tokens returned from deposit LP tokens and as rewards form onsen LP stake, we should get in return xSushi tokens
+     3- Calculate gains/loss
+     */
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        // reads
+        uint256 index = tokenIds[id];
+        IERC20 LPtoken = _tokens[index];
+        uint256 balance = LPtoken.balanceOf(address(this));
+
+        // interactions
+        //Deposit LP token into onsen, if deposit succesfull this address should have in return sushi tokens
+        if (balance > 0) {
+            IERC20(LPtoken).approve(masterChef, balance);
+            IMasterChef(masterChef).deposit(id, balance); // deposit into Onsen
+        }
+
+        // Stake the sushi tokens
+        if (IERC20(sushi).balanceOf(address(this)) > 0) {
+            enterSushiBar(true); // manage rewards
+        }
+
+        //Calculate gains/loss
+        // Retrieve current balance for pool and address
+        UserInfo memory currentUserInfo = IMasterChef(masterChef).userInfo(id, address(this));
+
+        uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+        if (currentUserInfo.amount >= last) {
+            gain = uint128(currentUserInfo.amount - last);
+        } else {
+            loss = uint128(last - currentUserInfo.amount);
+        }
+    }
+
+    /**
+        Asuming the index of the amounts [] can be use as parameter for the mapping tokenIds
+        Example:
+        amounts => [20,50,150]
+        So the values that will be used to retrieve the id used in onsen will be 0, 1 , 2
+    */
+    function deallocate(uint256[] memory amounts) public override onlyGuardian {
+        for (uint256 i; i < amounts.length; i++) {
+            uint256 amount = amounts[i];
+            if (amount > 0) {
+                IMasterChef(masterChef).withdraw(tokenIds[i], amount); // withdraw from Onsen
+            }
+        }
+    }
+
+    function _deactivate(bool panic) internal override {}
+
+    function _prepareMigration() internal override {}
```

Comment:
shouldn't this deallocate everything?

Reply:
@Paul4912 Yes it does, but I am having an issue trying to get the list of tokenIds to be deallocated, since 
  // From deposit id to the token's id
    mapping(uint256 => uint256) public tokenIds;
    
 is a mapping structure so I cannot iterate over it, and I haven't been able how to figure out which are the active ids from that list that are needed to call the withdraw function. 
 @Paul4912 / @ind-igo / @mirru2532  If you have any ideas please let me know, I will continue trying to figure out a way but so far I don't know.

Reply:
You need to hold an iterable array instead of a mapping, tokenIds is as said not for this usecase. Only transforming as mentioned above.